### ISSUE TITLE
testsuite: coll/nonblocking[23].c assume MPI_Datatype is an int

### ISSUE THREAD

**Issue by jeffhammond** [pmodels/mpich#6398]
`MPI_Datatype` is required to be a C `int`.

I will PR this later.  You can assign to me (I can't assign to myself).

```patch
mpich/test/mpi/coll$ git diff  .
diff --git a/test/mpi/coll/nonblocking2.c b/test/mpi/coll/nonblocking2.c
index 615d3082e..d05e39be6 100644
--- a/test/mpi/coll/nonblocking2.c
+++ b/test/mpi/coll/nonblocking2.c
@@ -58,8 +58,8 @@ int *sendcounts = NULL;
 int *recvcounts = NULL;
 int *sdispls = NULL;
 int *rdispls = NULL;
-int *sendtypes = NULL;
-int *recvtypes = NULL;
+MPI_Datatype *sendtypes = NULL;
+MPI_Datatype *recvtypes = NULL;
 signed char *buf_alias = NULL;
 MPI_Request req;

diff --git a/test/mpi/coll/nonblocking3.c b/test/mpi/coll/nonblocking3.c
index 52e20a034..ab652569f 100644
--- a/test/mpi/coll/nonblocking3.c
+++ b/test/mpi/coll/nonblocking3.c
@@ -95,8 +95,8 @@ struct laundry {
     int *recvcounts;
     int *sdispls;
     int *rdispls;
-    int *sendtypes;
-    int *recvtypes;
+    MPI_Datatype *sendtypes;
+    MPI_Datatype *recvtypes;
 };

 static void cleanup_laundry(struct laundry *l)
@@ -136,8 +136,8 @@ static void start_random_nonblocking(MPI_Comm comm, unsigned int rndnum, MPI_Req
jhammond@nuclear:~/mukautuva/mpich/test/mpi/coll$ git diff  .
diff --git a/test/mpi/coll/nonblocking2.c b/test/mpi/coll/nonblocking2.c
index 615d3082e..d05e39be6 100644
--- a/test/mpi/coll/nonblocking2.c
+++ b/test/mpi/coll/nonblocking2.c
@@ -58,8 +58,8 @@ int *sendcounts = NULL;
 int *recvcounts = NULL;
 int *sdispls = NULL;
 int *rdispls = NULL;
-int *sendtypes = NULL;
-int *recvtypes = NULL;
+MPI_Datatype *sendtypes = NULL;
+MPI_Datatype *recvtypes = NULL;
 signed char *buf_alias = NULL;
 MPI_Request req;

diff --git a/test/mpi/coll/nonblocking3.c b/test/mpi/coll/nonblocking3.c
index 52e20a034..ab652569f 100644
--- a/test/mpi/coll/nonblocking3.c
+++ b/test/mpi/coll/nonblocking3.c
@@ -95,8 +95,8 @@ struct laundry {
     int *recvcounts;
     int *sdispls;
     int *rdispls;
-    int *sendtypes;
-    int *recvtypes;
+    MPI_Datatype *sendtypes;
+    MPI_Datatype *recvtypes;
 };

 static void cleanup_laundry(struct laundry *l)
@@ -136,8 +136,8 @@ static void start_random_nonblocking(MPI_Comm comm, unsigned int rndnum, MPI_Req
     int *recvcounts = NULL;
     int *sdispls = NULL;
     int *rdispls = NULL;
-    int *sendtypes = NULL;
-    int *recvtypes = NULL;
+    MPI_Datatype *sendtypes = NULL;
+    MPI_Datatype *recvtypes = NULL;
     signed char *buf_alias = NULL;

     MPI_Comm_rank(comm, &rank);
```

**Comment by jeffhammond** [comment]
This is fixed by https://github.com/pmodels/mpich/commit/a7cddae9186ec91f4b4712dfd99ed299d2ba95fb so maybe someone can cherry pick that.

**Comment by hzhou** [comment]
Fixed by #6414
